### PR TITLE
With auto_brace_complete enabled, selected text now gets wrapped by braces

### LINKED ISSUE #56581: `auto_brace_complete` does not wrap selected text with the desired characters anymore (unlike `3.x`)
### Godot version

v4.0.dev.custom_build [897afb69a]

### System information

Linux Mint - Vulkan API 1.2.131 - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce GTX 1080

### Issue description

When trying to wrap selected text in quotes or parenthesis (or any other character of that nature) it will not wrap the selected text in the desired characters but instead replace the selected text with the pressed character.

This behavior occurs regardless of the value of the editor setting `auto_brace_complete`.

image

### Steps to reproduce

1. Select any text in the code editor
2. Press `(`

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by IgorKordiukiewicz**
Fixes #56581

Now it behaves the same as in 3.x

### CODE REVIEW

**Comment by Paulb23** [comment]
Any chance you could add some unit tests for this?

**Comment by IgorKordiukiewicz** [comment]
Added

**Review by Paulb23 — COMMENTED** [review]
Could you add a case for braces larger then one char? In the test it should be open / close key of `'''`

**Comment by IgorKordiukiewicz** [comment]
Shouldn't it only work for one char braces? Because the `CodeEdit::_handle_unicode_input_internal(const uint32_t p_unicode)` accepts a single char so I think it's impossible to wrap selection with multi char braces. For example if you would try to wrap it with the `/*` and `*/` pair, upon entering the `/` character, it would replace the selection. It also works that way in 3.x

**Comment by Paulb23** [comment]
Yeah I don't expect it to pass, though nice to see the negative case for it. 

Though in 4.0 outside a selection it should complete when you type the `*` in `/*`.

**Comment by IgorKordiukiewicz** [comment]
> Yeah I don't expect it to pass, though nice to see the negative case for it.

So should I just add something like this
```
/* If inserted character is not a auto brace completion open key, replace selected text with the inserted character */
		code_edit->clear();
		code_edit->insert_text_at_caret("abc");
		code_edit->select_all();
		SEND_GUI_KEY_EVENT(code_edit, Key::KEY_1);
		CHECK(code_edit->get_text() == "1");
```
but with e.g the `/` character instead of `1`?

> Though in 4.0 outside a selection it should complete when you type the * in /*.

Yeah it already works that way

**Comment by Paulb23** [comment]
Going to guess it should be something like this:
```
/* If inserted character is multichar brace completion open key, it does not wrap */
code_edit->clear();
code_edit->insert_text_at_caret("\'\'abc");
code_edit->select(0, 2, 0, 5);
SEND_GUI_KEY_EVENT(code_edit, `\'`);
CHECK(code_edit->get_text() == "\'\'\'");
```

**Comment by IgorKordiukiewicz** [comment]
Ok so it turns out it can actually easily work also with multichar braces. So if in line `''abc` , `abc` is selected and `'` pressed should the behaviour be: `'''abc'''` or `'''` ?

**Comment by Paulb23** [comment]
There doesn't seem to be much precedents for this in other editors, as it could also autocomplete to `''''''`. 

But for me personally, I would expect `'''`

**Comment by IgorKordiukiewicz** [comment]
Updated

File: tests/scene/test_code_edit.h [hunk 1]
```diff
@@ -2786,6 +2786,41 @@ TEST_CASE("[SceneTree][CodeEdit] completion") {
...
+		code_edit->select_all();
+		SEND_GUI_KEY_EVENT(code_edit, Key::KEY_1);
+		CHECK(code_edit->get_text() == "1");
+
+		/* If inserted character is a multichar brace completion open key, it does not wrap or complete */
+		code_edit->clear();
+		code_edit->insert_text_at_caret("\'\'abc");
+		code_edit->select(0, 2, 0, 5);
+		SEND_GUI_KEY_EVENT(code_edit, Key::APOSTROPHE);
+		CHECK(code_edit->get_text() == "\'\'\'");
```
Comment (Paulb23) [inline tests/scene/test_code_edit.h:2823]: There's another in test here, currently both `' '` and `''' '''` are registered as brace completions. When the user press the `'` in this scenario we should get: `'''acb'` as it should match `' '` not `''' '''`. If we remove `' '` from the `CodeEdit` then this behaviour looks good.

Comment (IgorKordiukiewicz) [reply tests/scene/test_code_edit.h:2823]: So basically this test should actually be `CHECK(code_edit->get_text() == "\'\'\'abc\'");` ?

Comment (Paulb23) [reply tests/scene/test_code_edit.h:2823]: Should be something like:

```c++
/* If potential multichar and single brace completion is matched, it should wrap the single.  */
code_edit->clear();
code_edit->insert_text_at_caret("\'\'abc");
code_edit->select(0, 2, 0, 5);
SEND_GUI_KEY_EVENT(code_edit, Key::APOSTROPHE);
CHECK(code_edit->get_text() == "\'\'\'abc\'");

/* If only the potential multichar brace completion is matched, it does not wrap or complete. */
auto_brace_completion_pairs.erase("\'");
code_edit->set_auto_brace_completion_pairs(auto_brace_completion_pairs);
CHECK_FALSE(code_edit->has_auto_brace_completion_open_key("\'"));

code_edit->clear();
code_edit->insert_text_at_caret("\'\'abc");
code_edit->select(0, 2, 0, 5);
SEND_GUI_KEY_EVENT(code_edit, Key::APOSTROPHE);
CHECK(code_edit->get_text() == "\'\'\'");
```

**Comment by IgorKordiukiewicz** [comment]
Updated

File: scene/gui/code_edit.cpp [hunk 2]
```diff
@@ -591,27 +593,40 @@ void CodeEdit::_handle_unicode_input_internal(const uint32_t p_unicode) {
...
+		bool try_to_wrap_selection_with_braces = (had_selection && auto_brace_completion_enabled);
 
-		if (has_string_delimiter(chr) && cc > 0 && _is_char(get_line(cl)[cc - 1]) && post_brace_pair == -1) {
-			insert_text_at_caret(chr);
-		} else if (cc < get_line(cl).length() && _is_char(get_line(cl)[cc])) {
-			insert_text_at_caret(chr);
-		} else if (post_brace_pair != -1 && auto_brace_completion_pairs[post_brace_pair].close_key[0] == chr[0]) {
-			caret_move_offset = auto_brace_completion_pairs[post_brace_pair].close_key.length();
-		} else if (is_in_comment(cl, cc) != -1 || (is_in_string(cl, cc) != -1 && has_string_delimiter(chr))) {
+		if (try_to_wrap_selection_with_braces) {
```
Comment (Paulb23) [inline scene/gui/code_edit.cpp:599]: `auto_brace_completion_enabled` will always be true here, can simplify to `if (had_selection) {`

**Review by Paulb23 — COMMENTED** [review]
One minor nitpick otherwise looks good.

File: scene/gui/code_edit.cpp (hunk 2 above)
Comment (IgorKordiukiewicz) [reply scene/gui/code_edit.cpp]: Updated

**Review by Paulb23 — APPROVED** [review]

**Pull request merged**